server/e2e_test.go

```diff
@@ -0,0 +1,279 @@
+package server
+
+import (
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"net"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cashshuffle/cashshuffle/message"
+)
+
+var (
+	testTempKey int
+)
+
+func TestHappyShuffle(t *testing.T) {
+	poolSize := 3
+	h := testNewHarness(t, poolSize)
+	bch1 := uint64(100000000)
+	version := uint64(999)
+
+	// clients join the pool, one at a time
+	clients := make([]*testClient, 0)
+	for i := 0; i < poolSize; i++ {
+		client := h.NewClient()
+		clients = append(clients, client)
+
+		client.Register(bch1, version)
+		if i < poolSize-1 {
+			h.AssertBroadcastNewPlayer(client, clients)
+		}
+	}
+	// the pool is full and the shuffle starts
+	h.AssertBroadcastPhase1Announcement(clients)
+
+	// confirm basic state on the server side
+	h.AssertPools([]poolState{
+		{
+			value:   bch1,
+			version: version,
+			players: len(clients),
+			isFull:  true,
+		},
+	})
+
+	// the shuffle succeeded, and clients leave with no blame
+	for _, c := range clients {
+		err := c.conn.Close()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// after the clients leave, the pool should be removed
+	h.AssertPools([]poolState{})
+
+	// all messages should be consumed through the assertions
+	// if anything is remaining, it is outside of specification
+	// or something unexpected happened
+	h.AssertEmptyInboxes(clients)
+}
+
+// testHarness holds the pieces required for automating a shuffle
+type testHarness struct {
+	tracker *Tracker
+	packets chan *packetInfo
+	t       *testing.T
+}
+
+type testClient struct {
+	h         *testHarness
+	conn      net.Conn
+	inbox     *testInbox
+	amount    uint64
+	version   uint64
+	session   []byte
+	playerNum uint32
+}
+
+// new TestHarness sets up the required parts for automating a shuffle
+func testNewHarness(t *testing.T, poolSize int) *testHarness {
```

Comment:
I think this should be `newTestHarness` =)

Reply:
Yeah... what do you think after our offline discussion?

1. pollute `server` namespace with ambiguously test-related stuff `newTest...`
2. or have slightly odd names that clearly isolate the test-related stuff `testNew...`
3. or make a new package for e2e testing and do all the necessary refactoring of `server`.

`3.` seems a bit too much to me. `2.` seemed the best compromise to me when I wrote it but maybe `1.` is better? Any other solutions?

Reply:
I think polluting in a small way and allowing `Test`, `newTest`, and `assert` prefixed methods is reasonable. Also the assert methods below I think should be unexported and start with `assert`. =)

Reply:
Sounds good. Should be done by b17bd34